src/ripple/consensus/LedgerTrie.h

```diff
@@ -0,0 +1,712 @@
+                {
+                    cs.append(child->getJson());
+                }
+            }
+            return res;
+        }
+    };
+
+    // The root of the trie. The root is allowed to break the no-single child
+    // invariant.
+    std::unique_ptr<Node> root;
+
+    /** Find the node in the trie that represents the longest common ancestry
+        with the given ledger.
+
+        @return Pair of the found node and the sequence number of the first
+                ledger difference.
+    */
+    std::pair<Node*, Seq>
+    find(Ledger const& ledger) const
+    {
+        Node* curr = root.get();
+
+        // Root is always defined and is in common with all ledgers
+        assert(curr);
+        Seq pos = curr->span.diff(ledger);
+
+        bool done = false;
+
+        // Continue searching for a better span as long as the current position
+        // matches the entire span
+        while (!done && pos == curr->span.end())
+        {
+            done = true;
+            // Find the child with the longest ancestry match
+            for (std::unique_ptr<Node> const& child : curr->children)
+            {
+                auto childPos = child->span.diff(ledger);
+                if (childPos > pos)
+                {
+                    done = false;
+                    pos = childPos;
+                    curr = child.get();
+                    break;
+                }
+            }
+        }
+        return std::make_pair(curr, pos);
+    }
+
+    void
+    dumpImpl(std::ostream& o, std::unique_ptr<Node> const& curr, int offset)
+        const
+    {
+        if (curr)
+        {
+            if (offset > 0)
+                o << std::setw(offset) << "|-";
+
+            std::stringstream ss;
+            ss << *curr;
+            o << ss.str() << std::endl;
+            for (std::unique_ptr<Node> const& child : curr->children)
+                dumpImpl(o, child, offset + 1 + ss.str().size() + 2);
+        }
+    }
+
+public:
+    LedgerTrie() : root{std::make_unique<Node>()}
+    {
+    }
+
+    /** Insert and/or increment the support for the given ledger.
+
+        @param ledger A ledger and its ancestry
+        @param count The count of support for this ledger
+     */
+    void
+    insert(Ledger const& ledger, std::uint32_t count = 1)
+    {
+        Node* loc;
+        Seq diffSeq;
+        std::tie(loc, diffSeq) = find(ledger);
+
+        // There is always a place to insert
+        assert(loc);
+
+        Span lTmp{ledger};
+        Span prefix = lTmp.before(diffSeq);
+        Span oldSuffix = loc->span.from(diffSeq);
+        Span newSuffix = lTmp.from(diffSeq);
+        Node* incNode = loc;
+
+        if (!oldSuffix.empty())
+        {
+            // new is a prefix of current
+            // e.g. abcdef->..., adding abcd
+            //    becomes abcd->ef->...
+
+            // Create oldSuffix node that takes over loc
+            std::unique_ptr<Node> newNode{std::make_unique<Node>(oldSuffix)};
```

Comment:
Don't need to say the type twice: `auto newNode = std::make_unique<Node>(oldSuffix)`.